runner/engine.go

```diff
@@ -583,9 +601,31 @@ func (e *Engine) runBin() error {
 	return nil
 }
 
+func (e *Engine) stopBin() {
+	e.mainDebug("initiating shutdown sequence")
+	start := time.Now()
+	e.mainDebug("shutdown completed in %v", time.Since(start))
+
+	exitCode := make(chan int)
+
+	e.withLock(func() {
+		if e.binStopCh != nil {
+			e.mainDebug("sending shutdown command to killfunc")
+			e.binStopCh <- exitCode
+			e.binStopCh = nil
+		} else {
+			close(exitCode)
+		}
+	})
+	if ret := <-exitCode; ret != 0 {
+		os.Exit(ret)
```

Comment:
Your refactoring is great, I love it.

I would have appreciated if this could be tested.

One solution for this is to define an exiter in the engine.

The exiter could be assigned to os.Exit by default, and overloaded via the tests to return a t.Fatal if the code is not the one expected

What do you think about it?

Reply:
looks good i will raise a pr for that.

Reply:
Thanks

Reply:
https://github.com/air-verse/air/pull/729